src/wallet/ethereum.ts

```diff
@@ -126,24 +127,50 @@ export async function init({ onAccountChange, onDisconnect }: InitArgs): Promise
 
   const ethereum = await load()
 
+  // accountsChanged is called when the account connected to the app changes - this includes when
+  // an additional account is connected, as well as when the connected account is disconnected,
+  // which will return an empty accounts array
   ethereum.on("accountsChanged", async (accounts: string[]) => {
-    handleAccountsChanged(accounts)
-    await onAccountChange()
+    if (accounts.length) {
+      // MetaMask can sometimes trigger accountsChanged when first connecting to an account, so we need to
+      // ensure it is actually being triggered by a new account change to avoid extra signatures
+      if (globCurrentAccount !== accounts[0]) {
+        handleAccountsChanged(accounts)
+        await onAccountChange()
+      }
+    } else {
+      // disconnected
+      await disconnect({ onDisconnect })
+    }
   })
 
+  // The MetaMask provider emits this event if it becomes unable to submit RPC requests to any chain.
+  // In general, this will only happen due to network connectivity issues or some unforeseen error.
   ethereum.on("disconnect", async () => {
-    globCurrentAccount = null
-    await onDisconnect()
+    await disconnect({ onDisconnect })
   })
 
   didBindEvents = true
 }
 
 
+async function disconnect({
+  onDisconnect,
+}: {
+  onDisconnect: () => Promise<unknown>
+}): Promise<void> {
+  globCurrentAccount = null
+  await onDisconnect()
+}
+
+
 export async function sign(data: Uint8Array): Promise<Uint8Array> {
+  const key = `${globCurrentAccount}-${data.toString()}`
+  if (globSignature[key]) return globSignature[key]
```

Comment:
I think you may be right. I added it early into my investigation because I saw a lot of places where `sign(MSG_TO_SIGN)` was being called and `MSG_TO_SIGN` was always referring to the same message, but it turned out I didn't need it to fix this flow. 

One place I thought it might be useful was for the [verifyPublicKey](https://github.com/fission-codes/webnative-walletauth/blob/avivash/multiple-metamask-popups/src/wallet/ethereum.ts#L382-L388) function, which will request two of the same signatures if `globPublicSignatureKey` isn't defined yet. But it also doesn't look `verifyPublicKey` is being called anywhere at the moment, so I'm happy to remove it. Though maybe we do in fact want to call `verifyPublicKey` somewhere? 😅

Reply:
`verifyPublicKey` only exists because I wanted to verify if the public key being generated was actually valid (so for testing purposes only).